x/emissions/module/network_summaries.go

```diff
@@ -0,0 +1,271 @@
+	maxPreviousRegret := float32(epsilon) // averts div by 0 error
+	for _, regret := range regrets.WorkerRegrets {
+		// If inference regret is not null, use it to calculate the maximum regret
+		if maxPreviousRegret < regret.InferenceRegret && float64(regret.InferenceRegret) > epsilon {
+			maxPreviousRegret = regret.InferenceRegret
+		}
+		// If forecast regret is not null, use it to calculate the maximum regret
+		if maxPreviousRegret < regret.ForecastRegret && float64(regret.ForecastRegret) > epsilon {
+			maxPreviousRegret = regret.ForecastRegret
+		}
+	}
+
+	// Calculate the network combined inference and network worker regrets
+	unnormalizedI_i := float64(0)
+	sumWeights := 0.0
+	for _, regret := range regrets.WorkerRegrets {
+		// normalize worker regret then calculate gradient => weight per worker for network combined inference
+		weight, err := gradient(pInferenceSynthesis, float64(regret.InferenceRegret/maxPreviousRegret))
+		if err != nil {
+			fmt.Println("Error calculating gradient: ", err)
+			return float64(0), err
+		}
+		unnormalizedI_i += weight * float64(inferenceByWorker[regret.Worker].Value.Uint64()) // numerator of network combined inference calculation
+		sumWeights += weight
+	}
+
+	// Normalize the network combined inference
+	if sumWeights < epsilon {
+		return 0, emissions.ErrSumWeightsLessThanEta
+	}
+	return unnormalizedI_i / sumWeights, nil // divide numerator by denominator to get network combined inference
+}
+
+// Gathers inferences, forecast-implied inferences as of the given block
+// and the network regrets admitted by workers at or just before the given time,
+// then invokes calculation of network combined inference I_i, network per worker regret R_i-1,l, and weights w_il from the litepaper.
+func GetAndCalcNetworkCombinedInference(
+	ctx sdk.Context,
+	k keeper.Keeper,
+	topicId TopicId,
+	blockHeight BlockHeight) (float64, error) {
+	// Get inferences from worker update at or just after the given block
+	inferences, err := k.GetInferencesAtOrAfterBlock(ctx, topicId, blockHeight)
+	if err != nil {
+		fmt.Println("Error getting inferences: ", err)
+		return float64(0), err
+	}
+	// Get regrets admitted by workers just before the given block
+	regrets, err := k.GetNetworkRegretsAtOrBeforeBlock(ctx, topicId, blockHeight)
+	if err != nil {
+		fmt.Println("Error getting network regrets: ", err)
+		return float64(0), err
+	}
+	// Get forecast-implied inferences at the given block
+	forecastImpliedInferences, err := GetAndCalcForcastImpliedInferencesAtBlock(ctx, k, topicId, blockHeight)
+	if err != nil {
+		fmt.Println("Error getting forecast implied inferences: ", err)
+		return float64(0), err
+	}
+
+	epsilon, err := k.GetParamsEpsilon(ctx)
+	if err != nil {
+		fmt.Println("Error getting epsilon: ", err)
+		return float64(0), err
+	}
+
+	pInferenceSynthesis, err := k.GetParamsPInferenceSynthesis(ctx)
+	if err != nil {
+		fmt.Println("Error getting epsilon: ", err)
+		return float64(0), err
```

Comment:
why this `float64(0)` in all returns? I can see other examples above that doesn't require converting 0 to float64

Reply:
not kenny, but usually depends on whether the compiler is smart enough to figure out the type by itself

but could also be written `return 0.0, err`